src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -111,7 +111,23 @@ public function validate($entity, Constraint $constraint)
             }
         }
 
-        $repository = $em->getRepository(get_class($entity));
+        if (null !== $constraint->repository) {
+            /* Retrieve repository from given entity name.
+             * We ensure the retrieved repository can handle the entity
+             * by checking the entity is the same, or subclass of the supported entity.
+             */
+            $repository = $em->getRepository($constraint->repository);
+            $supportedClass = $repository->getClassName();
+
+            $className = $class->getName();
+
+            if (!($supportedClass === $className || is_subclass_of($className, $supportedClass))) {
```

Comment:
This can be simplified with ``$entity instanceof $supportedClass``, as you don't only have the class name but an instance.